Allow Users to Revisit Check-in Meetings

nickoferrall opened the pull request:
This PR resolves issue #3704 

### Acceptance Criteria

- [ ] after a phase is complete it says "Phase Completed"
- [ ] after a meeting is completed, it says "Meeting Completed" in the left nav header
- [ ] can go to the timeline view & see the ended meeting. It includes the total tasks & comments & agenda item discussed
- [ ] from the timeline event can go back into an old meeting to see the discussion

Note, after ending a meeting and going to the timeline view, the user needs to refresh the page to be able to see the most recent meeting. This is a known issue #3682

mattkrick commented on packages/server/graphql/types/ActionMeeting.ts:
```diff
@@ -16,6 +17,43 @@ const ActionMeeting = new GraphQLObjectType<IActionMeeting, GQLContext>({
+      description: 'The number of comments generated in the meeting',
+      resolve: async ({commentCount}) => {
+        // only populated after the meeting has been completed (not killed)
+        return commentCount || 0
```
for older checkin meeetings, this is always going to return 0, which may be a bummer.
could you write a migration to determine how many comments & agenda items those? should look similar to https://github.com/ParabolInc/parabol/blob/master/packages/server/database/migrations/20200323143644-retroevents.ts#L21

nickoferrall replied:
ahh true! I've just added a migration 👍

mattkrick commented:
code looks good! testing now...

mattkrick commented on packages/client/components/TimelineEventCompletedActionMeeting.tsx:
```diff
@@ -46,9 +54,14 @@ class TimelineEventCompletedActionMeeting extends Component<Props> {
+          <CountItem>{`${commentCount} ${plural(commentCount, 'comment')}.`}</CountItem>
           <br />
-          <Link to={`/new-summary/${meetingId}`}>See the full summary</Link>
+          <Link to={`/meet/${meetingId}/updates/1`}>See the discussion</Link>
```
we should redirect to the discussion, not the updates

mattkrick commented:
it seems that clicking the agenda items doesn't navigate to the discussion thread around that agenda item.

AC
- [ ] clicking "See the discussion" goes to the first agenda item
- [ ] once in the ended meeting, clicking through the agenda items in the left nav allows you to see the threaded discussions

nickoferrall commented:
Hey @mattkrick , I’ve got the functionality working but I’m having difficulties getting TypeScript working with Relay. In `AgendaList.tsx`, for example, I’m passing `meeting` and `agendaItems` down as props. To get the type, I’m querying the meeting  and defining the meeting type as AgendaList_meeting. It feel unnecessary to create a query just to get the type so I’m not sure if I’m taking the right approach. 
For `agendaItems`, I’ve temporarily used `any` as I’m modifying the original type so when I use the same query approach, TypeScript complains about props being different. I’ve been Googling and I also tried creating an interface but as `AgendaItem.tsx` is using the query for the type, this doesn’t feel right.
Perhaps this would be a good one to discuss in our 1-1 tomorrow if there’s time.

mattkrick commented on packages/client/components/ActionSidebarAgendaItemsSection.tsx:
```diff
@@ -26,9 +26,19 @@ const ActionSidebarAgendaItemsSection = (props: Props) => {
+  const agendaItemsPhase = meeting.phases!.find(
+    (phase) => phase.phaseType === NewMeetingPhaseTypeEnum.agendaitems
+  )!
+  const agendaItems = useMemo(() => {
```
this component isn't directly using `agendaItems`, it's only passing it down along with the `meeting` parent object. may be cleaner to just pass down `meeting` and let the child component get what it needs

mattkrick commented on packages/client/components/ActionMeetingAgendaItems.tsx:
```diff
@@ -91,8 +89,18 @@ const ActionMeetingAgendaItems = (props: Props) => {
 graphql`
-  fragment ActionMeetingAgendaItemsStage on AgendaItemsStage {
-    agendaItemId
+  fragment ActionMeetingAgendaItemsStage on NewMeetingStage {
```
curious why you'd use a less specific type here?

mattkrick commented on packages/client/components/ActionSidebarAgendaItemsSection.tsx:
```diff
@@ -53,6 +64,14 @@ graphql`
         isComplete
         isNavigable
+        isNavigableByFacilitator
+        agendaItem {
```
can remove?

mattkrick commented on packages/client/components/DiscussionThread.tsx:
```diff
@@ -82,6 +82,17 @@ graphql`
   }
 `
+
+graphql`
```
is this used anywhere?

mattkrick commented on packages/client/modules/teamDashboard/components/AgendaAndTasks/AgendaAndTasks.tsx:
```diff
@@ -130,7 +130,11 @@ const AgendaAndTasks = (props: Props) => {
             </AgendaHeader>
-            <AgendaListAndInput dashSearch={dashSearch || ''} team={team!} />
+            <AgendaListAndInput
+              agendaItems={agendaItems}
```
you're already passing down team, and agendaItems is a sub-object of team, so no need to pass them both down

mattkrick commented on packages/client/modules/teamDashboard/components/AgendaList/AgendaList.tsx:
```diff
@@ -124,18 +125,12 @@ const AgendaList = (props: Props) => {
-      activeMeetings {
-        ...AgendaItem_activeMeetings
-        id
+  meeting: graphql`
```
if the component isn't directly using a property, it shouldn't be requesting it.
e.g. `endedAt` only appears once in this file-- in the graphql query. therefore, it shouldn't exist at all

nickoferrall commented:
Hey @mattkrick, I've cleaned up the queries and updated the types. It's ready for a re-review when you get the chance

mattkrick commented on packages/client/components/ActionMeetingUpdates.tsx:
```diff
@@ -93,6 +98,11 @@ export default createFragmentContainer(ActionMeetingUpdates, {
       id
       endedAt
       showSidebar
+      localPhase {
```
this is a big gotcha in relay.
we set up `localPhase` to be local state
local state is never fetched from the server. you can think of `localPhase` as a memoization of the available domain data.
so to guarantee that the local state includes `stages.isComplete`, we need to also request it from the server.

while this works today, it only works because another component is requesting `isComplete`. if that component no longer requests that field, it won't be available here.

so, on L111 you'll need to add `isComplete`. that way it guarantees it'll be available within the local object.

nickoferrall replied:
ah very interesting! I've just pushed an update including `isComplete`

mattkrick commented on packages/client/modules/teamDashboard/components/AgendaListAndInput/AgendaListAndInput.tsx:
```diff
@@ -31,30 +35,70 @@ interface Props {
-  meetingId?: string
+}
+
+const getAgendaItems = (meeting) => {
```
can add the typing of `AgendaListAndInput_meeting | null` to meeting so you don't need the `meeting.phases!`

mattkrick commented on packages/client/modules/teamDashboard/components/AgendaListAndInput/AgendaListAndInput.tsx:
```diff
@@ -31,30 +35,70 @@ interface Props {
-  const {dashSearch, gotoStageId, isDisabled, team, meetingId} = props
+  const {dashSearch, gotoStageId, isDisabled, team, meeting} = props
+  const endedAt = meeting?.endedAt
+  const agendaItems = endedAt ? getAgendaItems(meeting) : team.agendaItems
```
while they should be the same, it might be more expressive to say `getAgendaItems(meeting) || team.agendaItems`.

if there's a meeting, grab the agenda from the meeting. otherwise, grab it from the team

mattkrick commented:
- In completed meetings, the "Add Agenda Items" link should probably be hidden.
- same goes for the pro tip (Psst) if there are no agenda items
- The lastcall phase could include different copy (e.g. There were no agenda items in this meeting...)



![Screen Shot 2020-08-31 at 6 08 53 PM](https://user-images.githubusercontent.com/5514175/91783428-101a7a00-ebb5-11ea-88d0-3f8fdce3c78d.png)

mattkrick commented on packages/server/database/migrations/20200807100000-addCheckInMetrics.ts:
```diff
@@ -0,0 +1,54 @@
+        (row) => ({
+          commentCount: r
+            .table('Comment')
+            .getAll(getThreadIds(row), {index: 'threadId'})
```
this migration doesn't work. you're passing in a comment, but on L10 you're using the `meetingId` as an index.
to test, make sure that on the timeline the agenda item count & the comment count return a non-zero for some of the completed meetings.

nickoferrall replied:
Hmm I might be mistaken here but my understanding is that in the `.update` we're looping through each `NewMeeting` and passing a `meeting` into the `getThreadIds`, not a `comment`? That function returns the agenda items ids which we use to grab the comments.

I've updated the migration removing the active filter and how `agendaItemCount` was calculated and I'm now getting non-zero counts for completed meetings.

nickoferrall commented:
Hey @mattkrick, I’ve made those updates and also changed the copy for the first call phase.

I haven’t re-added the delete agenda items functionality as I was thinking that completed meetings should be read-only rather than editable. If we’re disabling the ability to add new items, perhaps it shouldn’t be possible to remove or edit them either? This would be consistent with the completed retro meetings where it’s not possible to delete a topic.

With that in mind, I’ve disabled the DnD in the agenda items sidebar, disabled the comment / task input and buttons and updated the empty thread copy for both retro & check ins.

I can certainly revert the changes and re-add the delete functionality. I was just thinking that completed meetings should be either read-only or editable whereas I feel they’re currently read-only in terms of the inability to add items but editable in the sense that users can add comments.

mattkrick commented on packages/client/components/DiscussionThreadInput.tsx:
```diff
@@ -137,13 +148,19 @@ const DiscussionThreadInput = forwardRef((props: Props, ref: any) => {
         editorRef={editorRef}
         editorState={editorState}
         onSubmit={onSubmit}
+        readOnly={!!endedAt}
```
@ackernaut should discussion threads close after the meeting is complete?

ackernaut replied:
If it’s easy enough I’m okay with it. If it’s complicated now or later I’d want to spend more time thinking about the async user stories between sync time, like when folks might pick a deeper discussion back up later, etc.

mattkrick commented:
Almost there!

## Issue 1: Cannot delete agenda items for active meetings
Reproduction:
1. start a new checkin meeting
2. advance to the updates phase
3. add 3 agenda items
4. hover the 2nd agenda item & click the X to delete it. notice that it does not go away

## Issue 2: Merge conflicts
let's wait on @ackernaut to confirm if disabling the discussion thread is OK after the meeting is ended. I can't remember what the final verdict on this was

ackernaut commented:
As of now, folks have little way of knowing if others are still using the thread, but there’s little harm in keeping them open. In a fuzzy future, there may be a way where folks can save a conversation for later or we may notify folks about active threads. But the user stories and IA need to be worked out.

If the Retro threads stay active, I’d say let’s keep these active. If it’s simple as flipping a `boolean` then it’s safe whatever we do (2-way decision). If it’s something complicated then I’d say keep it open as it has been I assume until we figure out better UX.

nickoferrall commented on packages/client/components/DiscussionThread.tsx:
```diff
@@ -28,16 +30,18 @@ const Wrapper = styled('div')<{isExpanded: boolean}>(({isExpanded}) => ({
-  const {commentors, thread} = threadSource!
-  const threadSourceId = threadSource!.id!
-  const preferredNames = commentors && commentors.map((commentor) => commentor.preferredName)
+  const thread = threadSource?.thread
```
I came across a strange bug where `threadSource` was null and an error was thrown. I ended a check-in meeting, clicked into it from the Timeline and went to the Discuss stage where the bug was thrown. I've created several different accounts and tried many combinations of ending a meeting but I haven't been able to repro it.

Line 45 mentions an unusual Relay error relating to `thread` that Terry reported. Perhaps this is the same issue or perhaps something strange happened with the gql executor? If you login with user: t1@lisboa.com pass: testing you can find the bug in Check-in #5 on the first agenda item but not the subsequent 3. The `threadSourceId` is invalid. 

To prevent the app from crashing I made this change of line 41 and pass the `threadSourceId` down from the parent but it's not possible to add a comment/task to this agenda item.

nickoferrall commented:
OK cool, makes sense @ackernaut. I've updated this PR to remain consistent with how things currently work, i.e. it is possible to add comments and tasks in completed retros and check-ins but it's not possible to add new agenda items, DnD or remove items.

@mattkrick thanks for all your patience on this PR! Those two issues are now resolved so it's ready for a re-review when you have the time
